### ISSUE TITLE
implement common subresources

### ISSUE THREAD

**Issue by clux** [clux/kube-rs#127]
E.g. if anyone needs special case subresource functionality like:

- [ ] pods/attach
- [ ] pods/portforward
- [ ] pods/eviction
- [ ] pods/exec
- [x] pods/log (already implemented, used as example herein)
- [ ] service/proxy
- [ ] node/proxy

We have a decent pattern for subresources this that doesn't require massive amounts of generics:

1. Implement subresource queries straight on `RawApi` without restrictions (for now):

```rust
impl RawApi {
    /// Get a pod logs
    pub fn log(&self, name: &str, lp: &LogParams) -> Result<http::Request<Vec<u8>>> {
        ...
        // Check https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/ for urls
    }
}
```

2. Make a marker trait for the subresource:

```rust
pub trait LoggingObject {}

impl<K> Api<K>
where
    K: Clone + DeserializeOwned + KubeObject + LoggingObject,
{
    pub async fn log(&self, name: &str, lp: &LogParams) -> Result<String> {
        let req = self.api.log(name, lp)?;
        Ok(self.client.request_text(req).await?)
    }

    pub async fn log_stream(&self, name: &str, lp: &LogParams) -> Result<impl Stream<Item = Result<Bytes>>> {
        let req = self.api.log(name, lp)?;
        Ok(self.client.request_text_stream(req).await?)
    }
}
```

3. Designate what resources implement this subresource:

```rust
impl LoggingObject for Object<PodSpec, PodStatus> {}
impl LoggingObject for Object<PodSpec, Void> {}
impl LoggingObject for Object<DeploymentSpec, DeploymentStatus> {}
impl LoggingObject for Object<DeploymentSpec, Void> {}
```

That ought to be it. Write one test case in examples (e.g. [examples/log_openapi.rs](https://github.com/clux/kube-rs/blob/master/examples/log_openapi.rs)).

These are easy PRs to do for new contributors who need extra functionality for subresources!

**Comment by edio** [comment]
Hi, I'd like to give this a try (I'm interested in support for `exec` mostly and would like to contribute what I can (which is not much at this point) to bring its support to kube-rs sooner).

I do understand, that this task is just about implementing subresources in rust code, not about handling them correctly. However one item is particularly confusing to me

> 1. Implement subresource queries straight on Resource without restrictions (for now)

I started with the `logs` example and came up with this for the `exec`

```
    pub fn exec(&self, name: &str, ep: &ExecParams) -> Result<http:Request<Vec<u8>>> {
        let base_url = self.make_url() + "/" + name + "/" + "exec?";
        let mut qp = url::form_urlencoded::Serializer::new(base_url);

        if let Some(container) = &lp.container {
            qp.append_pair("container", &container);
        }

        ...

        let req = http::Request::post(urlstr);
        req.body(vec![]).map_err(Error::HttpError)
    }
```

The issue is that `exec` can't be performed over http. It used to require SPDY (which is now deprecated) and now relies on websockets. Hence, an attempt to send a http request to this endpoint will yield `400` as response. 

I wanted to clarify, whether you deliberately have chosen `Result<http::Request<...>>` as a return type to serve as a placeholder or you had something else in mind here.

If it's the former, should we maybe resort to `Any` to prevent folks using kube-rs from relying on these parts of API while full support is not there yet (you can't do much with `Any` I imagine)?

**Comment by clux** [comment]
> The issue is that `exec` can't be performed over http. It used to require SPDY (which is now deprecated) and now relies on websockets. Hence, an attempt to send a http request to this endpoint will yield `400` as response.
> 
> I wanted to clarify, whether you deliberately have chosen `Result<http::Request<...>>` as a return type to serve as a placeholder or you had something else in mind here.
> 
> If it's the former, should we maybe resort to `Any` to prevent folks using kube-rs from relying on these parts of API while full support is not there yet (you can't do much with `Any` I imagine)?

Honestly, I wasn't aware of how much harder subresources like `exec` or port-forward` were going to be. We have a bit of an open discussion about it in #229, but we haven't gotten to a satisfying conclusion there yet.

Would assume that we need another return type for `exec`. `Any` doesn't sound very useful from an api reuse pov, nor from a user pov. It almost certainly needs a new transport dependency and a new signature specifically for the subresource.

That said, I haven't thought too much about this yet. If you want to help sketch out something or discuss on #229 that would be much appreciated.

**Comment by clux** [comment]
`0.46.0` introduces the `ws` feature using `async-tungstenite` to pave the way for the last subresources listed above. Have updated the issue slightly. Implementations of the last ws-based subresources can look at #360 for an excellent PR that covers (among the websocket handling itself) examples for implementing `pods/attach` and `pods/exec`.

**Comment by kazk** [comment]
I'll do `pods/portforward`. kube should probably just provide `AsyncRead + AsyncWrite` for each port and not bind to local ports. Users should be able to do that if they want to.

I have a _rough_ sketch that can do something similar to [Python client's example](https://github.com/kubernetes-client/python/blob/818e5ba0ba8cfc6d866195de706f58b7431110c1/examples/pod_portforward.py#L97-L117):
```rust
// Portforward nginx container
let mut pf = pods.portforward("example", &[80]).await?; // Portforwarder
let mut ports = pf.ports().unwrap(); // Vec<Port>
let mut port = ports[0].stream().unwrap(); // impl AsyncRead + AsyncWrite + Unpin
port.write_all(b"GET / HTTP/1.1\r\nHost: 127.0.0.1\r\nConnection: close\r\nAccept: */*\r\n\r\n")
    .await?;
let mut rstream = tokio_util::io::ReaderStream::new(port);
if let Some(res) = rstream.next().await {
    match res {
        Ok(bytes) => {
            let response = std::str::from_utf8(&bytes[..]).unwrap();
            println!("{}", response);
            assert!(response.contains("Welcome to nginx!"));
        }
        Err(err) => eprintln!("{:?}", err),
    }
}
```

I'll open a draft PR later (it's too messy at the moment to show...) to discuss the design.

That should make kube a gold client except for "Proto encoding".

> Gold Requirements [![Client Capabilities](https://camo.githubusercontent.com/67dd557ef9144e5acfe3b401cd59cd514077ebf794d092d54a460b724bbad3b4/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f4b756265726e65746573253230636c69656e742d476f6c642d626c75652e7376673f7374796c653d706c617374696326636f6c6f72423d46464437303026636f6c6f72413d333036434538)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/csi-new-client-library-procedure.md#client-capabilities)
> 
> - Support exec, attach, port-forward calls (these are not normally supported out of the box from [swagger-codegen](https://github.com/swagger-api/swagger-codegen))    
> - Proto encoding

I'm not sure what that means. Do we need to support Protobufs?

**Comment by clux** [comment]
Yeah, that sounds sensible. Listening sounds like an application concern. In an example we can maybe use `TcpListener` or something and direct it to the writable stream to show how to do the `kubectl port-forward` equivalent.

> protobufs

Ugh. Ah, yeah, I am not sure that's really possible atm. The go api has protobuf  codegen hints (see [api/types.go](https://github.com/kubernetes/api/blob/master/core/v1/types.go) like:

```go
// +optional
// +patchMergeKey=name
// +patchStrategy=merge
EphemeralContainers []EphemeralContainer `json:"ephemeralContainers,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,34,rep,name=ephemeralContainers"`
```

whereas  the swagger codegen has the equivalent json for that part of the PodSpec:

```json

**Comment by kazk** [comment]
The official JavaScript client is Gold and just have something like this (only deserializing):

```typescript
export class ProtoClient {
    public readonly 'config': KubeConfig;
    public async get(msgType: any, requestPath: string): Promise<any> {
        // ...
        const req = http.request(options);
        const result = await new Promise<any>((resolve, reject) => {
            let data = '';
            req.on('data', (chunk) => {
                data = data + chunk;
            });
            req.on('end', () => {
                const obj = msgType.deserializeBinary(data);
                resolve(obj);
            });
            req.on('error', (err) => {
                reject(err);
            });
        });
        req.end();
        return result;
    }
}
```
https://github.com/kubernetes-client/javascript/blob/master/src/proto-client.ts

(BTW, it can't portforward more than one port. https://github.com/kubernetes-client/javascript/blob/a8d4f1c7bea2b27403d380e22e492b6680f80b31/src/portforward.ts#L19)

I guess the requirement for Gold is not well-defined? Python one seems more capable, but that's Silver.

---

When I searched "Kubernetes Proto Encoding", https://kubernetes.io/docs/reference/using-api/api-concepts/#protobuf-encoding came up.

**Comment by clux** [comment]
Yeah, I think the client grading document is a thing they they wrote and applied to the clients they generated themselves. It technically doesn't apply to us (because we don't have our client in that place they say it should be in), but it's a nice benchmark otherwise. I have opened an issue to continue protobuf discussions in #371 at any rate.

**Comment by chroche** [comment]
Naive question about `pods/eviction`, this feature doesn't seem to get much love at all. Is there a possible workaround / a direct way to create ak8s_openapi [Eviction](https://docs.rs/k8s-openapi/0.11.0/k8s_openapi/api/policy/v1beta1/struct.Eviction.html)? Is there a reason why there seems to be no interest in this feaure?

**Comment by clux** [comment]
It's probably just that no one has asked for it yet. It actually looks very simple. I can have a go at  adding it.

**Comment by clux** [comment]
Eviction handled in #393

**Comment by jmintb** [comment]
Can I have a go at implementing ephemeral container support?

**Comment by clux** [comment]
@jmintb Go for it!

**Comment by jmintb** [comment]
Alright so as far as I can tell we want to support replacing, updating and reading ephemeral containers for a pod. I have found the following operations in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#podspec-v1-core.
`PATCH /api/v1/namespaces/{namespace}/pods/{name}/ephemeralcontainers`
`GET /api/v1/namespaces/{namespace}/pods/{name}/ephemeralcontainers`
`PUT /api/v1/namespaces/{namespace}/pods/{name}/ephemeralcontainers`

This also matches what the JS client does.

Am I missing anything?

**Comment by clux** [comment]
That looks right to me!

One minor point; we probably have to put this behind an `k8s_openapi::k8s_if_ge_1_25` guard when importing the structs since afaikt it only stabilised in 1.25.

**Comment by jmintb** [comment]
Awesome, that is my understanding as well.

**Comment by jmintb** [comment]
I noticed while writing tests for the ephemeral containers, that I can submit an invalid patch without an error occurring, should that be the case or have I broken something? 

Maybe that is just how the JSON merge patch works.

Examples:

Bad case, where the patch is only the `Pod` spec and not the entire `Pod` object. I would expect an error here but get none.
```rust
let patch = serde_json::json!(
           {
           "ephemeralContainers": [
           {
               "name": "myephemeralcontainer",
               "image": "busybox:1.34.1",
               "command": ["sh", "-c", "sleep 20"],
           },
           ]
        });

        pod = pods
            .patch_ephemeral_containers(
                &pod.name_any(),
                &PatchParams::default(),
                &Patch::Merge(patch.clone()),
            )
            .await?;

```

Valid case with a partial `Pod` object.
```rust
let patch: Pod = serde_json::from_value(json!({
            "spec":{
            "ephemeralContainers": [
            {
                "name": "myephemeralcontainer",
                "image": "busybox:1.34.1",
                "command": ["sh", "-c", "sleep 20"],
            },
            ]
        }}))?;

        pod = pods
            .patch_ephemeral_containers(
                &pod.name_any(),
                &PatchParams::default(),
                &Patch::Merge(patch.clone()),
            )
            .await?;
```

**Comment by bestgopher** [comment]
> "ephemeralContainers": {
>           "description": "...",
>           "items": {
>             "$ref": "#/definitions/io.k8s.api.core.v1.EphemeralContainer"
>           },
>           "type": "array",
>           "x-kubernetes-patch-merge-key": "name",
>           "x-kubernetes-patch-strategy": "merge"
>         },

@clux how to add x-kubernetes-patch-merge-key and x-kubernetes-patch-strategy for custom resource?

**Comment by clux** [comment]
> > "ephemeralContainers": {
> > "description": "...",
> > "items": {
> > "$ref": "#/definitions/io.k8s.api.core.v1.EphemeralContainer"
> > },
> > "type": "array",
> > "x-kubernetes-patch-merge-key": "name",
> > "x-kubernetes-patch-strategy": "merge"
> > },
> 
> [@clux](https://github.com/clux) how to add x-kubernetes-patch-merge-key and x-kubernetes-patch-strategy for custom resource?

That's currently not supported. I wasn't even aware it existed.

I guess it's specifically for strategic merge?  tried looking around for it a bit, and found the [kubernetes page](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch-using-the-retainkeys-strategy) on strategic merge, but in the wild the results are older. Is it possible that strategic merge has gone out of fashion in favor of kubernetes recommended [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) and the [new merge strategies](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) are sufficient?

If it is still useful, I'm not opposed to adding it. Please open issue (this issue is unrelated)! Presumably it would be a feature request for the `KubeSchema` derive so you could do `#[x_kube(patch ... something)]`.

<details><summary>misc stuff i found interesting about this</summary>
<p>

1. [core schemas](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/customOpenAPIschema.md) (suspect because they can do more in the schemas than crds can)
2. old blog post about [alertmanager config](https://tech.aabouzaid.com/2022/11/set-openapi-patch-strategy-for-kubernetes-custom-resources-kustomize.html) which are now only using `x-kubernetes-map-type: atomic`
3. some wip tooling openapi tooling in https://github.com/paperclip-rs/paperclip/blob/437bd84645fcc46656a4020ccba3d3d71b45cc43/src/v2/mod.rs#L24 by Tiago.
4. [crds not touched in 2 years](https://github.com/lyft/flinkk8soperator/blob/9eae44173ddf1dee9f8bbf3263100a4bc07a3049/deploy/crd.yaml#L405)

unfortunately did not find anything comparing patch strategies to merge strategies.
</p>
</details>
